Initial MCH sim implementation

Juliacare opened the pull request.

Juliacare committed:
Add files via upload

Juliacare committed:
mch 2

Juliacare committed:
Merge branch 'xiv-gear-planner:master' into master

Juliacare committed:
Merge branch 'xiv-gear-planner:master' into master

VioletHynes committed:
[MCH Sim] Get the skeleton working

Juliacare committed:
Merge pull request #1 from VioletHynes/violet/mch-skeleton

[MCH Sim] Get the skeleton working

Juliacare committed:
Merge branch 'xiv-gear-planner:master' into master

Juliacare committed:
MCH setup 3

- GCD APL
- oGCD APL
- Queen usage
todo:
forcenocrit, forcenodhit on WF

Juliacare committed:
Force no DH/ no Crit

Juliacare committed:
Update xivstats.ts

return 0 for forcenocrit/dh instead

Juliacare committed:
add forcenocrit/dh to sim types

Juliacare commented:
theres a mistake somewhere in the rotation handling. Its doing funky stuff right now.
![image](https://github.com/user-attachments/assets/20029542-d4da-444a-9d11-63fd5389e013)
![image](https://github.com/user-attachments/assets/2fd9e557-332f-47e8-94e7-d476c0cb6d21)

Juliacare committed:
fix blazing shot usage

xiashtra commented on package.json:
```diff
@@ -16,14 +16,16 @@
     "@stylistic/eslint-plugin-js": "^2.9.0",
     "@types/eslint": "^9.6.1",
+    "beasties": "^0.3.3",
+    "beasties-webpack-plugin": "^0.3.3",
     "eslint": "^9.13.0",
     "eslint-plugin-chai-friendly": "^1.0.1",
     "eslint-plugin-named-import-spacing": "^1.0.3",
     "http-server": "^14.1.1",
     "nyc": "^17.1.0",
     "swagger-typescript-api": "^13.0.22",
-    "tsconfig-paths-webpack-plugin": "^4.2.0",
     "ts-node": "^10.9.2",
+    "tsconfig-paths-webpack-plugin": "^4.2.0",
```
This file shouldn't be getting modified by your sim PR

Juliacare replied:
seems like it got pushed through by mistake. Got modified from building locally

xiashtra commented:
You should run the lint script; there's a lot of lint errors that's making it slow scan what's actually wrong.

Juliacare committed:
Fix some more stuff

Juliacare committed:
lint stuff

Juliacare commented:
> You should run the lint script `npm run lint`; there's a lot of lint errors that's making it slow scan what's actually wrong.

https://github.com/xiv-gear-planner/gear-planner/pull/556/commits/16b5e6b1cd32f28410d8f9aeaf0ce49f0c33d697 
should be done in here. 2 errors remain:
```E:\gearcalc\gear-planner\packages\core\src\sims\ranged\mch\mch_sheet_sim.ts
  109:15  warning  'buffs' is assigned a value but never used. Allowed unused vars must match /^_$/u  @typescript-eslint/no-unused-vars
  423:13  error    'mchAbility' is never reassigned. Use 'const' instead                              prefer-const```
not sure what to do with those

xiashtra commented on packages/core/src/sims/ranged/mch/mch_sheet_sim.ts:
```diff
@@ -0,0 +1,512 @@
+            return null;
+        }
+
+        let mchAbility = ability as MchAbility;
```
```suggestion
        const mchAbility = ability as MchAbility;
```
fixes the lint error

Juliacare committed:
full metal field naming thing

Juliacare commented:
main issues right now are:
- it doesnt appear like hypercharge is being used after the opener, could be caused by other issues? no idea
- im not sure if queen is actually doing damage, it is being used though and battery is being reset accordingly
- barrel stabalizer gives a buff that allows the usage of full metal field but this buff doesnt go away after usage so its leading to the sim spamming the action for the entire 30s duration of the buff

xiashtra commented on packages/core/src/sims/ranged/mch/mch_sheet_sim.ts:
```diff
@@ -0,0 +1,512 @@
+        this.advanceTo(this.currentTime + (remainingtime - totalAnimLock));
+    }
+
+    applyAutomatonQueenAbility(abilityUsage: AutomatonQueenAbilityUsageTime) {
+        const buffs = [...this.getActiveBuffs(abilityUsage.usageTime)];
+    }
```
This doesn't do anything?  The lint error is because you don't use the `buffs` constant for anything once you set it, but the whole function is a NOP right now.

xiashtra commented on packages/core/src/sims/ranged/mch/mch_sheet_sim.ts:
```diff
@@ -0,0 +1,512 @@
+        cp.AutomatonQueenAbilityUsages.push({
+            // Pile Bunker
+            ability: {
+                ...Actions.AutomatonQueenArmPunch,
```
```suggestion
                ...Actions.AutomatonQueenPileBunker,
```

xiashtra commented on packages/core/src/sims/ranged/mch/mch_sheet_sim.ts:
```diff
@@ -0,0 +1,512 @@
+        cp.AutomatonQueenAbilityUsages.push({
+            // Crowned Collider
+            ability: {
+                ...Actions.AutomatonQueenArmPunch,
```
```suggestion
                ...Actions.AutomatonQueenCrownedCollider,
```

xiashtra reviewed (commented):
You have 7 Arm Punches instead of using Pile Bunker & Crowned Collider at the end

xiashtra commented on packages/core/src/sims/ranged/mch/mch_types.ts:
```diff
@@ -0,0 +1,117 @@
+export type MchExtraData = {
+    /** The battery gauge data */
+    gauge: MchGaugeState,
+    hypercharge: number,
```
I don't think you actually need this anywhere?  It's extra data for the UI but you're not displaying it currently.

xiashtra commented on packages/core/src/sims/ranged/mch/mch_types.ts:
```diff
@@ -0,0 +1,117 @@
+};
+
+export const FullMetalFieldBuff: Buff = {
+    name: "Full Metal Field",
```
```suggestion
    name: "Full Metal Machinist",
```
buff name from game data

xiashtra commented on packages/core/src/sims/ranged/mch/mch_types.ts:
```diff
@@ -0,0 +1,117 @@
+};
+
+
+export const HyperchargeBuff: Buff = {
+    name: "Hypercharge",
+    duration: 10,
+    selfOnly: true,
+    stacks: 5,
+    effects: {
+        // Adds potency to attacks.
+    },
+    appliesTo: ability => ability.type === "gcd",
+    beforeAbility<X extends MchAbility>(buffController: BuffController, ability: X): X {
+        return {
+            ...ability,
+            potency: ability.potency + 20,
+        };
+    },
+    beforeSnapshot<X extends MchAbility>(buffController: BuffController, ability: X): X {
+        buffController.subtractStacksSelf(1);
+        return ability;
+    },
+    statusId: 688,
+};
```
```suggestion
export const OverheatedBuff: Buff = {
    name: "Overheated",
    duration: 10,
    selfOnly: true,
    stacks: 5,
    effects: {
        // Adds potency to attacks.
    },
    appliesTo: ability => ability.type === "gcd",
    beforeAbility<X extends MchAbility>(buffController: BuffController, ability: X): X {
        return {
            ...ability,
            potency: ability.potency + 20,
        };
    },
    beforeSnapshot<X extends MchAbility>(buffController: BuffController, ability: X): X {
        buffController.subtractStacksSelf(1);
        return ability;
    },
    statusId: 2688,
};
```
This is the current status that Hypercharge gives.  688 is an older status from when it applied to turrets.

xiashtra commented on packages/core/src/sims/ranged/mch/mch_types.ts:
```diff
@@ -0,0 +1,117 @@
+};
+
+export const ReassembleBuff: Buff = {
+    name: "Reassemble",
```
```suggestion
    name: "Reassembled",
```
buff name from game data

xiashtra committed:
Update package.json

xiashtra committed:
initial lint

xiashtra committed:
yasss queen

xiashtra commented:
> * im not sure if queen is actually doing damage, it is being used though and battery is being reset accordingly

progress
![image](https://github.com/user-attachments/assets/9234edaa-5c72-4cfe-9427-5ac0c9b351ba)

xiashtra committed:
hypercharge fix, blazing shot fix

- Fixed Hypercharge to not spend Heat when the Hypercharged buff is present
- Fixed Blazing Shot not reducing the cooldown of Double Check and Checkmate
- some formatting to improve code readability

Juliacare committed:
Merge pull request #2 from xiashtra/mch-debug

MCH sim fixes

Juliacare commented:
I got a comment from MCH TC. They'd like the gauges to be coloured correctly. Is that easy to do? Violet set up the UI so im not sure.

```const HEAT_COLOR = Color('#D35A10')
const BATTERY_COLOR = Color('#2C9FCB') ```

xiashtra committed:
additional automaton queen fixes

xiashtra committed:
Update mch_sheet_sim.ts

xiashtra committed:
fix hypercharge not being used

xiashtra commented:
> I got a comment from MCH TC. They'd like the gauges to be coloured correctly. Is that easy to do? Violet set up the UI so im not sure.
> 
> const HEAT_COLOR = Color('#D35A10') const BATTERY_COLOR = Color('#2C9FCB')

Violet set it up so you could pick a color for each gauge, and also a secondary color when the gauge reaches a certain threshold (e.g., you could have the Heat Gauge be orange from 0-49, and red from 50-100 to indicate when Hypercharge is available).

hintxiv commented on packages/frontend/src/scripts/sims/ranged/mch_sheet_sim_ui.ts:
```diff
@@ -0,0 +1,119 @@
+
+                    const barInner = document.createElement('div');
+                    // @TODO Juliacare: pick a colour, and pick a threshold for it to be a different colour
+                    barInner.style.backgroundColor = gauge >= 50 ? '#ffc500' : '#e47e08';
```
@Juliacare here's where you can modify the gauge colour (likewise for battery). I'm not sure what you want to do for the "greater than threshold" colour, up to you.

xpdota replied:
These could also be CSS styles in `guage.less`

hintxiv commented on packages/core/src/sims/ranged/mch/mch_actions.ts:
```diff
@@ -0,0 +1,242 @@
+export const AutomatonQueenArmPunch: MchOgcdAbility = {
+    type: 'ogcd',
+    name: "(Automaton Queen) Arm Punch",
+    alternativeScalings: ["Pet Action Weapon Damage"],
```
We need another queen-specific scaling to account for her special dex mod (100 instead of the player's 115). You can reference how Living Shadow is handled for this.

We also need to strip party bonus from her dex stat, DRK seems to handle this for Living Shadow specifically but it would probably be beneficial to add this to the generic pet action handling since there are no pets which benefit from party bonus.

xiashtra replied:
This should already be handled by the changes @VioletHynes made to support pet actions: https://github.com/xiv-gear-planner/gear-planner/blob/483e6d2e86f4020dbe8a45a7973ce843a9e7c8d0/packages/xivmath/src/xivmath.ts#L156-L159

Juliacare committed:
Merge pull request #3 from xiashtra/mch-debug

more MCH fixes

hintxiv commented on packages/core/src/sims/ranged/mch/mch_actions.ts:
```diff
@@ -0,0 +1,242 @@
+    potency: 240 * 6,
+    // 240 per GCD, 6 hits
+    // 1440 potency total
+    attackType: "Ability",
```
Wildfire is technically a DoT and therefore uses a slightly different damage formula than a direct damaging ability. Is this being handled?

xiashtra replied:
Not currently.  It could be faked by treating it as a DoT with one tick.

hintxiv commented on packages/core/src/sims/ranged/mch/mch_sheet_sim.ts:
```diff
@@ -0,0 +1,584 @@
+    getTimeUntilDoubleCheckCapped(): number {
+        return this.cdTracker.statusOf(Actions.DoubleCheck).cappedAt.relative;
+    }
+    // @TODO Juliacare: this function isn't being used anywhere currently
```
You should go through and clean up all of these TODO comments, I think they mostly don't apply anymore.

xiashtra replied:
All of those apply; there are a lot of functions in the sim that aren't ever called for anything currently.

xpdota commented on packages/core/src/sims/ranged/mch/mch_gauge.ts:
```diff
@@ -0,0 +1,45 @@
+
+export class MchGauge {
+
+    private _batteryGauge: number = 0;
```
I have a question: How is this supposed to be handled when you have a propery guarded by an accessor pair? If you just call it `batteryGauge`, then it conflicts with `get batteryGauge`/`set batteryGauge`

hintxiv replied:
@xpdota usually there's a more human readable way of naming the variables, e.g.

```ts
private battery = 0
public get currentBattery() {
    return this.battery
}
```

xpdota commented on packages/core/src/sims/ranged/mch/mch_sheet_sim.ts:
```diff
@@ -0,0 +1,569 @@
+    usageTime: number,
+}
+
+class RotationState {
```
I would say that it would also be nice to validate that '3' is actually the length of the array of combo actions.

xiashtra commented on packages/core/src/sims/ranged/mch/mch_sheet_sim.ts:
```diff
@@ -0,0 +1,569 @@
+    applyAutomatonQueenAbility(abilityUsage: AutomatonQueenAbilityUsageTime) {
+        const buffs = [...this.getActiveBuffs(abilityUsage.usageTime)];
+
+        // @TODO Juliacare: filter out buffs that do not apply to Automaton Queen's actions
```
It's less about raid buffs, and more a note to be careful about personal buffs not being consumed by Queen's actions.

Juliacare replied:
Good thing MCH doesnt have personal buffs. The only thing I could think off is HC but that only applies to GCDs and queen GCDs are treated as OGCDS

xiashtra commented on packages/core/src/sims/ranged/mch/mch_sheet_sim.ts:
```diff
@@ -0,0 +1,569 @@
+
+    useAutomatonQueen(cp: MchCycleProcessor) {
+        // After 5.6 delay, it does the following rotation with
+        // 2.5 seconds between each.
```
The code doesn't match the comment here; Arm Punch and Pile Bunker use `automatonQueenDelayBetweenAbilities = 1.5`, Crown Collider uses a 2.5s delay.

Juliacare replied:
Yea it doesnt. It's put at 1.5s delay by default and PB and CC have added time. 0.5 and 1.0s
![image](https://github.com/user-attachments/assets/5c012921-98de-412d-8992-f49ed865f740)
ill clarify it a little bit and fix the comment

Juliacare committed:
Merge branch 'xiv-gear-planner:master' into master

Juliacare committed:
some misc stuff

VioletStardustXIV commented on packages/core/src/sims/ranged/mch/mch_sheet_sim.ts:
```diff
@@ -362,7 +348,7 @@ export class MchSim extends BaseMultiCycleSim<MchSimResult, MchSettings, MchCycl
                     cp.cdTracker.canUse(Actions.AirAnchor, cp.nextGcdTime) ||
                     cp.cdTracker.canUse(Actions.Chainsaw, cp.nextGcdTime) ||
-                    cp.cdTracker.canUse(Actions.Excavator, cp.nextGcdTime)
+                    cp.isExcavatorReadyBuffActive()
```
This should also check if you can use Excavator without clipping, e.g. `cp.canUseWithoutClipping(Actions.Excavator)`

Juliacare replied:
isnt this part of the reassemble logic. Why does that matter?

VioletStardustXIV replied:
Oh, you're right, it doesn't in this case -- sorry, my lack of MCH knowledge :)

VioletStardustXIV commented on packages/core/src/sims/ranged/mch/mch_sheet_sim.ts:
```diff
@@ -387,19 +373,19 @@ export class MchSim extends BaseMultiCycleSim<MchSimResult, MchSettings, MchCycl
-        if (cp.remainingGcdTime <= 0) {
-            return;
+        if (cp.remainingGcdTime > 0) {
+            this.use(cp, this.getGCDToUse(cp));
```
and actually, would result in being unable to weave between some GCDs

Juliacare committed:
fix ogcd usage

Juliacare commented:
things fixed/added:
- shouldHypercharge implemented
- gauge colours
- more clear check to return early if no time for gcd is left
- PB delay .5 sec more than reg queen delay of 1s
- todo comments done
- excavator check now checks buff instead of cd

things still open:
- queen specific scaling
- style nit picks
- https://github.com/xiv-gear-planner/gear-planner/pull/556#discussion_r2054549476
- wildfire is a dot
- todo about `fightEndingsoon`, want to implement a correct way to use gauge at the end of a fight using this logic

Noted from current rotation:
- queen usage is a bit wonky. See images. It overcaps at 3:30 to 4:00 by a fair amount due to how we said it can use queen. This is because the queen at 2:00 is used at 80 battery instead of 100. Sounds wild but maybe we just hardcore queen usage in as a sequence of "use at these battery amount"
- it uses wildfire kind of delayed. I assume its because of excavator and the logic surrounding only using it during overheated.

![image](https://github.com/user-attachments/assets/cf5d79ec-09d9-438a-8cfc-636087e6028d)

zodiia commented:
Hello, was looking to potentially write a simulator for MCH as it's my main job and would have loved to be able to use that, and have enough knowledge to do so, as well as having the ability to maintain it with future job changes.

I see this has been open but no updates in 9 months, is there still plans on finishing it and merging it?

I'm willing to give a try at fixing what's left to fix here if the original author isn't active anymore but unsure if I would be able to do so or should rather start from scratch.

Thanks!
